backend/src/apiserver/server/pipeline_upload_server_test.go

```diff
@@ -534,22 +414,59 @@ func TestDefaultUpdatedPipelineVersion(t *testing.T) {
-	server = PipelineUploadServer{resourceManager: resourceManager, options: &PipelineUploadServerOptions{CollectMetrics: false}}
-	b = &bytes.Buffer{}
-	w = multipart.NewWriter(b)
-	part, _ = w.CreateFormFile("uploadfile", "arguments-version.tar.gz")
-	fileReader, _ = os.Open("test/arguments_tarball/arguments-version.tar.gz")
-	io.Copy(part, fileReader)
-	w.Close()
-	req, _ = http.NewRequest("POST", "/apis/v1beta1/pipelines/upload_version?pipelineid="+resource.DefaultFakeUUID, bytes.NewReader(b.Bytes()))
-	req.Header.Set("Content-Type", w.FormDataContentType())
-	rr = httptest.NewRecorder()
-	handler = http.HandlerFunc(server.UploadPipelineVersion)
-	handler.ServeHTTP(rr, req)
+	server = updateClientManager(clientManager, util.NewFakeUUIDGeneratorOrFatal(fakeVersionUUID, nil))
+	bytesBuffer, writer = setupWriter("")
+	setWriterFromFile("uploadfile", "arguments-version.tar.gz", "test/arguments_tarball/arguments-version.tar.gz", writer)
+	response = uploadPipeline(server, "POST", "/apis/v1beta1/pipelines/upload_version?pipelineid="+resource.DefaultFakeUUID,
+		bytes.NewReader(bytesBuffer.Bytes()), "Content-Type", writer.FormDataContentType(), server.UploadPipelineVersion)
+	assert.Equal(t, 200, response.Code)
 
 	pipeline, err := clientManager.PipelineStore().GetPipeline(resource.DefaultFakeUUID)
 	assert.Nil(t, err)
 	assert.Equal(t, pipeline.DefaultVersionId, fakeVersionUUID)
 }
+
+func setWriterWithBuffer(fieldname string, filename string, buffer string, writer *multipart.Writer) {
+	part, _ := writer.CreateFormFile(fieldname, filename)
+	io.Copy(part, bytes.NewBufferString(buffer))
+	writer.Close()
+}
+
+func setWriterFromFile(fieldname string, filename string, filepath string, writer *multipart.Writer) {
+	part, _ := writer.CreateFormFile(fieldname, filename)
+	fileReader, _ := os.Open(filepath)
+	io.Copy(part, fileReader)
+	writer.Close()
+}
+
+func setupWriter(text string) (*bytes.Buffer, *multipart.Writer) {
+	bytesBuffer := &bytes.Buffer{}
+	if text != "" {
+		bytesBuffer.WriteString(text)
+	}
+	return bytesBuffer, multipart.NewWriter(bytesBuffer)
+}
+
+func setupClientManagerAndServer() (*resource.FakeClientManager, PipelineUploadServer) {
+	clientManager := resource.NewFakeClientManagerOrFatal(util.NewFakeTimeForEpoch())
+	resourceManager := resource.NewResourceManager(clientManager)
+	server := PipelineUploadServer{resourceManager: resourceManager, options: &PipelineUploadServerOptions{CollectMetrics: false}}
+	return clientManager, server
+}
+
+func updateClientManager(clientManager *resource.FakeClientManager, uuid util.UUIDGeneratorInterface) PipelineUploadServer {
+	clientManager.UpdateUUID(uuid)
+	resourceManager := resource.NewResourceManager(clientManager)
+	server := PipelineUploadServer{resourceManager: resourceManager, options: &PipelineUploadServerOptions{CollectMetrics: false}}
+	return server
+}
+
+func uploadPipeline(server PipelineUploadServer, method string, url string, body io.Reader, requestKey string,
```

Comment:
Hi @NikeNano, thank you for the update!

I apologize, I just realized we could first iterate on `uploadPipeline` interface. When we finalize on the interface, you'll be able to finish the implementation in one go without back and forths with huge cost on your side.

Some parameters seems unnecessary:
Method will always be "POST", requestKey is always "Content-Type", requestValue always come from the same variable.

I think the conceptual way to look at whether this interface is a good fit is
 first thinking about what parameters `uploadPipeline(...)` needs. e.g. I think it needs the `uploadServer`, `pipelineName?`, `pipelineContent`, that's pretty much it. Then you can figure out what types they are.

After that, you can think about `uploadPipelineVersion(...)`, it needs `uploadServer`, `pipelineId`, `pipelineContent`, `pipelineVersionName`.

I understand some of the tests are testing an error case that you need special content in the request:
```
writer.CreateFormFile("uploadfile", "hello-world.yaml")
writer.Close()
```

So it might be a little hard to achieve the ideal interface, I totally understand that.

I think you've done a wonderful refactoring that greatly improved readability of these test cases.
Just get rid of the duplicative parameters and we are good to go.

Reply:
Thank you, appreciate all the feedback! I might look in to the other tests later and try to refactor them as well if needed.

Reply:
I update and removed the unnecessary parameters you pointed out from the uploadPipeline function.